common/types/string.go

```diff
@@ -221,3 +240,447 @@ func stringStartsWith(s String, pre ref.Val) ref.Val {
+	if err != nil {
+		return -1, "", NewErr("could not parse formatting clause: %s", err)
+	}
+
+	var argStr string
+	argAny := args.Get(Int(argIndex))
+	if formatter == nil {
+		return -1, "", NewErr("formatter required")
+	} else {
+		argStr, err = formatter(argAny, locale)
+		if err != nil {
+			return -1, "", NewErr("error during formatting: %s", err)
+		}
+	}
+	return i, argStr, nil
+}
+
+type clauseImpl func(ref.Val, string) (string, error)
+
+func formatString(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argStrVal := arg.ConvertToType(StringType)
+		argStr, ok := argStrVal.Value().(string)
+		if !ok {
+			return "", fmt.Errorf("could not convert argument %q to string", argStrVal)
+		}
+		return argStr, nil
+	}, nil
+}
+
+func formatFixed(lastStrIndex int, formatStr string, precision *int) (int, clauseImpl, error) {
+	if precision == nil {
+		return -1, nil, errors.New("precision required for fixed-point output")
+	}
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argFloatVal := arg.ConvertToType(DoubleType)
+		argFloat, ok := argFloatVal.Value().(float64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to float64", argFloatVal.Value())
+		}
+		fmtStr := fmt.Sprintf("%%.%df", *precision)
+		matchedLocale := matchLanguage(locale)
+		return message.NewPrinter(matchedLocale).Sprintf(fmtStr, argFloat), nil
+	}, nil
+}
+
+func formatScientific(lastStrIndex int, formatStr string, useUpper bool, precision *int) (int, clauseImpl, error) {
+	if precision == nil {
+		return -1, nil, errors.New("precision required for scientific output")
+	}
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argFloatVal := arg.ConvertToType(DoubleType)
+		argFloat, ok := argFloatVal.Value().(float64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to float64", argFloatVal.Value())
+		}
+		var fmtStr string
+		if useUpper {
+			fmtStr = fmt.Sprintf("%%%dE", *precision)
+		} else {
+			fmtStr = fmt.Sprintf("%%%de", *precision)
+		}
+		matchedLocale := matchLanguage(locale)
+		return message.NewPrinter(matchedLocale).Sprintf(fmtStr, argFloat), nil
+	}, nil
+}
+
+func formatBinary(lastStrIndex int, formatStr string) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argIntVal := arg.ConvertToType(IntType)
+		argInt, ok := argIntVal.Value().(int64)
+		if !ok {
+			return "", fmt.Errorf("could not convert \"%s\" to int64", argIntVal.Value())
+		}
+		// locale is intentionally unused as integers formatted as binary
+		// strings are locale-independent
+		return fmt.Sprintf("%b", argInt), nil
+	}, nil
+}
+
+func formatHex(lastStrIndex int, formatStr string, useUpper bool) (int, clauseImpl, error) {
+	return lastStrIndex + 1, func(arg ref.Val, locale string) (string, error) {
+		argIntVal := arg.ConvertToType(IntType)
```

Comment:
I was expecting Hex to also support string and byte sequences.

Reply:
Hmm, due to Go supporting this? (I wasn't aware that was possible before, very good to know) I think in terms of least-surprise I think it might be safer to keep it to just integers - I'm not aware of the hexadecimal formatting option doing this with other languages, but I could be wrong.

Reply:
Go does support this. Having a way to safely output byte strings is biggest motivator here. Developers rarely need to output a single int in hex, but it far more common that they need to safely output bytes.

Reply:
(marking as resolved since `formatHex` supports strings/byte buffers now)